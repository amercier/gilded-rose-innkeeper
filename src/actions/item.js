import {
  ITEMS_POLL_START,
  ITEMS_POLL_STOP,
  ITEMS_FETCH,
  ITEMS_FETCH_ERROR,
  ITEMS_SET,
  ITEMS_NAME_SEARCH,
  ITEMS_QUALITY_FILTER,
} from '../constants/actionTypes';
import {
  ITEMS_POLL_DELAY,
  ITEMS_SLIDER_FILTER_DEBOUNCE,
} from '../constants/config';

/**
 * Create an action to start polling items every ITEMS_POLL_DELAY ms, until a ITEMS_POLL_STOP action
 * is triggered.
 *
 * @returns {Object} An action that consists in starting polling items.
 */
export const doStartPollingItems = () => ({
  type: ITEMS_POLL_START,
  delay: ITEMS_POLL_DELAY,
});

/**
 * Create an action to stop polling items.
 *
 * @returns {Object} An action that consists in stopping polling items.
 */
export const doStopPollingItems = () => ({
  type: ITEMS_POLL_STOP,
});

/**
 * Create an action to add items.
 *
 * @param {Item[]} items - Items to add.
 * @returns {Object} An action that consists in adding the given items.
 */
export const doSetItems = items => ({
  type: ITEMS_SET,
  items,
});

/**
 * Create an action to fetch items.
 *
 * @returns {Object} An action that consists in fetching items.
 */
export const doFetchItems = () => ({
  type: ITEMS_FETCH,
});

/**
 * Create an action to handle an items fetch error.
 *
 * @param {Error} error - Error thrown during fetch.
 * @returns {Object} An action that consists in hnadling items fetch error.
 */
export const doFetchItemError = error => ({
  type: ITEMS_FETCH_ERROR,
  error,
});

/**
 * Create an action to update name search filter.
 *
 * @param {string} query - The new name search query.
 * @returns {Object} An action that consists in updating the name filter.
 */
export const doNameSearch = query => ({
  type: ITEMS_NAME_SEARCH,
  query,
});

/**
 * Create an action to update the quality min/max filter.
 *
 * @param {number} rangeStart - Minimum quality to display.
 * @param {number} rangeEnd - Maximum quality to display.
 * @returns {Object} An action that consists in updating the quality filter.
 */
export const doQualityFilter = (rangeStart, rangeEnd) => ({
  type: ITEMS_QUALITY_FILTER,
  rangeStart,
  rangeEnd,
  meta: {
    debounce: {
      time: ITEMS_SLIDER_FILTER_DEBOUNCE,
    },
  },
});
