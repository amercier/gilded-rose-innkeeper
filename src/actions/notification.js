import { NOTIFICATION } from '../constants/actionTypes';

/**
 * Create an action to display a notification.
 *
 * @param {string} message - Notification main message.
 * @param {string} description - Notification description.
 * @returns {Object} An action that consists in adding the given items.
 */
export const doNotification = (message, description) => ({
  type: NOTIFICATION,
  message,
  description,
});
