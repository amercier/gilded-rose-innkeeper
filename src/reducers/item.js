import {
  ITEMS_ADD,
  ITEMS_NAME_SEARCH,
  ITEMS_QUALITY_FILTER,
  ITEMS_SELL_IN_FILTER,
} from '../constants/actionTypes';

/**
 * @typedef Item
 *
 * Item.
 *
 * @type {Object}
 * @property {string} id - Item unique identifier.
 * @property {string} name - Item name.
 * @property {number} sellIn - Sell price.
 * @property {number} quality - Quality (between 0 and 100).
 * @property {string} type - Type of item: "STANDARD", "CONJURED", "BACKSTAGE_PASS" or "LEGENDARY".
 */

const INITIAL_STATE = {
  items: [],
  nameSearch: '',
  qualityMin: 0,
  qualityMax: 100,
  qualityRangeStart: null,
  qualityRangeEnd: null,
  sellInMin: null,
  sellInMax: null,
  sellInRangeStart: null,
  sellInRangeEnd: null,
};

/**
 * Get the minimum and maximum value of a given property within an array of objects.
 *
 * @param {Objects[]} objects - Objects.
 * @param {string} property - Proprty name.
 * @param {*} defaultMinValue - Value returned as min for an empty array.
 * @param {*} defaultMaxValue - Value returned as max for an empty array.
 * @returns {number[]} - An array containing the min an max value: `[min, max]`.
 */
function getMinMax(objects, property, defaultMinValue, defaultMaxValue) {
  return objects.reduce(
    ([min, max], { [property]: value }) => {
      const newMin = min === defaultMinValue ? value : Math.min(min, value);
      const newMax = max === defaultMaxValue ? value : Math.max(max, value);
      return [newMin, newMax];
    },
    [defaultMinValue, defaultMaxValue],
  );
}

/**
 * Add items and update sellIn min/max values.
 *
 * @param {Object} state - State.
 * @param {Item[]} items - Items to add.
 * @returns {Object} - The new state.
 */
function doItemsAdd(state, items) {
  const [sellInMin, sellInMax] = getMinMax(items, 'sellIn', null, null);
  return {
    ...state,
    items,
    sellInMin,
    sellInMax,
  };
}

/**
 * Item reducer.
 *
 * @param {Object} [state=INITIAL_STATE] - Previous state.
 * @param {string} action - Action type and data.
 * @returns {Object} Next state.
 */
function itemReducer(state = INITIAL_STATE, action) {
  switch (action.type) {
    case ITEMS_ADD: {
      return doItemsAdd(state, action.items);
    }
    case ITEMS_NAME_SEARCH: {
      return { ...state, nameSearch: action.query };
    }
    case ITEMS_QUALITY_FILTER: {
      return {
        ...state,
        qualityRangeStart: action.rangeStart,
        qualityRangeEnd: action.rangeEnd,
      };
    }
    case ITEMS_SELL_IN_FILTER: {
      return {
        ...state,
        sellInRangeStart: action.rangeStart,
        sellInRangeEnd: action.rangeEnd,
      };
    }
    default:
      return state;
  }
}

export default itemReducer;
