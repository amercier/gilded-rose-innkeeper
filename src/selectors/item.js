/**
 * Get a function that filters objects having a string property matching a given search query.
 *
 * @param {string} property - Property in the subject object.
 * @param {string} query - Text to search.
 * @returns {Function} - A function of the type: `(subject: Object) => boolean`, that returns `true`
 * if `subject[property]` contains `query`, or `false` otherwise.
 */
function propertySearchFilter(property, query) {
  const searchQuery = query.toLowerCase();
  return subject => subject[property].toLowerCase().includes(searchQuery);
}

/**
 * Get a function that filters objects having a number property within a given min/max range.
 *
 * @param {string} property - Property in the subject object.
 * @param {number} min - Minimum value.
 * @param {number} max - Maximum value.
 * @returns {Function} - A function of the type: `(subject: Object) => boolean`, that returns
 * `false` if `subject[property]` is lower than `min` or higher than `max`. Otherwise returns
 * `true`.
 */
function propertyMinMaxFilter(property, min, max) {
  return subject => subject[property] >= min && subject[property] <= max;
}

/**
 * Get a function that combine filters.
 *
 * @param {Function} filters - Filter functions of the type: `(subject: Object) => boolean`.
 * @returns {Function} - A function of the type: `(subject: Object) => boolean`, that returns `true`
 * if ALL filters return `true`, `false` if ANY of the filters returns `false`.
 */
function all(...filters) {
  return subject => filters.every(filter => filter(subject));
}

/**
 * Return all visible items.
 *
 * Right now all items are visible, so we just return all of them.
 *
 * @param {Object} state - Redux state.
 * @returns {Item[]} All visible items.
 */
function getVisibleItems({
  items,
  nameSearch,
  qualityRangeStart,
  qualityRangeEnd,
}) {
  return Object.values(items).filter(
    all(
      propertySearchFilter('name', nameSearch),
      propertyMinMaxFilter('quality', qualityRangeStart, qualityRangeEnd),
    ),
  );
}

export { getVisibleItems };
