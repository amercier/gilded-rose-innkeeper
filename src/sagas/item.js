import { delay } from 'redux-saga';
import { call, put, race, take, takeEvery, all } from 'redux-saga/effects';
import { doFetchItems, doSetItems } from '../actions/item';
import { API_URL_ITEMS } from '../constants/api';
import {
  ITEMS_POLL_START,
  ITEMS_POLL_STOP,
  ITEMS_FETCH,
} from '../constants/actionTypes';

/**
 * Fetch items from the API.
 *
 * @returns {Promise<Item[]>} All items.
 */
export async function fetchItems() {
  const response = await fetch(API_URL_ITEMS);
  return response.json();
}

/**
 * Generate two effects:
 * 1. A call to fetchItems
 * 2. A new ITEMS_SET actions with the items returned by fetchItems.
 *
 * @returns {Generator} A generator that yields a call to fetchItems and a ITEMS_SET action.
 */
export function* handleFetchItems() {
  const items = yield call(fetchItems);
  yield put(doSetItems(items));
}

/**
 * Generate the following effects infinitely:
 * 1. A new ITEMS_FETCH action.
 * 2. A call to wait() with the given delay.
 *
 * @param {number} ms - Time to wait before next poll, in ms.
 * @returns {Generator} A generator that yields infinitely a ITEMS_SET action and a call to wait.
 */
export function* handlePollItems(ms) {
  while (true) {
    yield put(doFetchItems());
    yield call(delay, ms);
  }
}

/**
 * Generate the following effects infinitely:
 * 1. A wait for the ITEMS_POLL_START action.
 * 2. A race between a handlePollItems call and a wait for ITEMS_POLL_STOP. This means effects from
 *    handlePollItems will be treated infinitely, until a ITEMS_POLL_STOP action is received.
 *
 * @returns {Generator} A generator that yields infinitely a wait forITEMS_POLL_START action and the
 * polling race.
 */
function* watchPollItemsStart() {
  while (true) {
    const action = yield take(ITEMS_POLL_START);
    yield race([call(handlePollItems, action.delay), take(ITEMS_POLL_STOP)]);
  }
}

/**
 * Create the items poll start/stop saga.
 *
 * @returns {Generator} The items poll start/stop saga.
 */
export default function* itemsSaga() {
  yield all([takeEvery(ITEMS_FETCH, handleFetchItems), watchPollItemsStart()]);
}
