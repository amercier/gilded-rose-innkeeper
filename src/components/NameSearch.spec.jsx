import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { Input } from 'antd';
import { doNameSearch } from '../actions/item';
import ConnectedNameSearch, { NameSearch } from './NameSearch';

describe('NameSearch', () => {
  it('renders its components', () => {
    const nameSearch = shallow(
      <NameSearch value="INITIAL VALUE" onChange={() => {}} />,
    );
    expect(nameSearch).toMatchSnapshot();
  });
});

describe('NameSearch (Redux-connected)', () => {
  const mockStore = configureStore();

  it('renders without crashing', () => {
    mount(
      <Provider store={mockStore({ nameSearch: '' })}>
        <ConnectedNameSearch />
      </Provider>,
    );
  });

  it('uses "nameSearch" store property as <Input> value', () => {
    const nameSearch = mount(
      <Provider store={mockStore({ nameSearch: 'INITIAL VALUE' })}>
        <ConnectedNameSearch />
      </Provider>,
    );
    const inputComponent = nameSearch.find(Input).first();
    expect(inputComponent.props().value).toBe('INITIAL VALUE');
  });

  it('triggers an "ITEMS_NAME_SEARCH" action on <Input> change', () => {
    const store = mockStore({ nameSearch: '' });
    const nameSearch = mount(
      <Provider store={store}>
        <ConnectedNameSearch />
      </Provider>,
    );
    const inputComponent = nameSearch.find(Input).first();
    const target = { value: 'NEW VALUE' };
    inputComponent.props().onChange({ target });
    expect(store.getActions()).toEqual([doNameSearch('NEW VALUE')]);
  });
});
