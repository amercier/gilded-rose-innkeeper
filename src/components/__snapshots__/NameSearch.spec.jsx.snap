// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameSearch renders its components 1`] = `
<Input
  disabled={false}
  onChange={[Function]}
  onClick={[Function]}
  placeholder="Name"
  prefix={
    <Icon
      style={
        Object {
          "color": "#1890ff",
        }
      }
      type="search"
    />
  }
  suffix={
    <Icon
      onClick={[Function]}
      style={
        Object {
          "color": "rgba(0,0,0,.25)",
        }
      }
      type="close"
    />
  }
  type="text"
  value="INITIAL VALUE"
/>
`;
